apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/MetricRegistry.java

```diff
@@ -172,6 +172,18 @@ public void report(MetricsReporter metricsReporter) {
         }
     }
 
+    public void resetBuffers() {
```

Comment:
This is called each time any event is processed from the queue when `server_urls` is empty, correct?
What is the purpose of flipping active and inactive metric sets in this case? Note it doesn't reset the counters and timers as that's only done during serialization. Conflating serialization and metrics resetting is probably a code smell and that's on me. But do we have to reset the metrics at all? If so, I think this also resets even if a span is reported, for example, which doesn't feel right. Or is this serving an entirely different purpose?

Documenting the purpose here would help.

Reply:
> This is called each time any event is processed from the queue when server_urls is empty, correct?

No, every time a metrics event is reported - https://github.com/elastic/apm-agent-java/pull/1295/files#diff-517f2e7d31c3425c5ec7719efd32a06dR151.
The purpose is to continue collecting metrics in the right intervals, because the `server_urls` is a dynamic config.

> Note it doesn't reset the counters and timers as that's only done during serialization.

Right. I was going to look into that next. What do you say simply replacing the buffers with new ones, as in:
```
    public void resetBuffers() {
        try {
            phaser.readerLock();
            inactiveMetricSets = activeMetricSets;
            activeMetricSets = new ConcurrentHashMap<>();
            phaser.flipPhase();
        } finally {
            phaser.readerUnlock();
        }
    }
```

> Documenting the purpose here would help.

👍

Reply:
> simply replacing the buffers with new ones

That won't work for gauges that still need to remain in the registry. Having to re-add transaction timers would also increase the allocation rate.

Reply:
@felixbarny check out the last commit. I separated reset from reporting. It costs a bit more because of the additional iteration, but since it is done as a secluded read-lock-operation, I think it's lightweight enough to worth the code simplification (especially now that we need to do that when not reporting as well).
I would even go further with providing a separate `resetState` operation for the `MetricRegistry`, but since it needs to be done on the same read-operation critical section with buffer switch and reporting, I didn't.

Reply:
Added tests as well